### DISCUSSION TITLE
Where the stubs are stored?

### DISCUSSION THREAD

**Discussion by fcasals** [BradenM/micropy-cli#352] [Q&A] [answered]
Hello! I'm new using micropy & micropython and it's been a nice journey so far.
I'm working with several venv's and I'm been installing micropy-cli on each one. So far, no fuzz.
In my first venv, I install a couple of stubs using `micropy stub add` command. Then, in my second venv, using `micropy init`, it offers me the same stubs installed in my first venv. So I'm assuming that micropy uses a common location to store the stubs, at OS/User level or local interpeter level. Could you help me to understant where are located? I'll like to add more stubs in that location.

**Comment by BradenM** [comment, marked as answer]
Hi @fcasals,

> So I'm assuming that micropy uses a common location to store the stubs, at OS/User level or local interpeter level. 

Your assumption is correct. `micropy` maintains a central local repository of stubs that is located under a `.micropy/stubs` folder that is created in the platform dependent user home directory[^1].

Generally speaking, this means:
- `~/.micropy/stubs/` on *nix platforms
- `c:\Users\MyUser\.micropy\stubs` on windows platforms.


When executing `micropy add`, micropy will:

1.  Ensure the stub exists in the central stubs repository (retrieving it if necessary).
2.  Check if the current working directory is a micropy-enabled project:

	* If it is a micropy project, also associate the stub with the project (creating links in the project's .micropy folder if needed)

	* Otherwise, do nothing. (so it only adds the stub for use, nothing more).


I'd be happy to clarify if needed. Thank you!


[^1]: This is the case as of the latest stable release at the time of writing (v4.0.0), at least. It is my intention to update this location to properly follow the standards most other applications align with. That is, following [XDG Base Spec](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html) env vars on *nix platforms, `~/Library/Application Support` on darwin/macos, and so on. This will be done in a way that maintains compatibility with the logic mentioned above.

**Reply by fcasals** [reply]
Thanks for your quick answer.
Cheers!
